### ISSUE TITLE
Auto focus is broken on New room page

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#17479]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Login with any account
2. Tab FAB menu > New room

## Expected Result: 
Auto focus on room name input and keyboard shows
## Actual Result: 
Room name input is not focused and "Create room" button bounces

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3..0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/232314025-3d269f41-d613-4604-a379-99dd9cadb0a7.MP4

https://user-images.githubusercontent.com/43996225/232314049-26306250-f0f1-4a49-81b9-1746731c39ac.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @situchan 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1681493406622809

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @johncschuster (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
@johncschuster Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by johncschuster** [comment]
I'm not sure how to "tab" the FAB menu on mobile. I've asked for clarity on the repro steps in the Slack thread above.

**Comment by MelvinBot** [comment]
@johncschuster Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by johncschuster** [comment]
I've reproduced the behavior. Triaging!

**Comment by MelvinBot** [comment]
Triggered auto assignment to @techievivek (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by johncschuster** [comment]
Umm... this isn't overdue, Melv.

**Comment by techievivek** [comment]
This can be worked externally.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~0160117b7c7096e58e

**Comment by MelvinBot** [comment]
Current assignee @johncschuster is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @thesahindia (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @PauloGasparSv (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by kuluruvineeth** [comment]
@thesahindia i am able to run web and desktop applications but while i am running npm run ios after running pod install in ios folder, i am not able to run ios(because of version compatibility issue) or android apps....according to local development environment setup there is SO link provided but looks like that is accessible only through expensify domain email which i dont have.....could you please provide further guidelines for me to setup and run ios or android app. Your assistance would really mean a lot for me to start contributing to this project, thanks.

**Comment by MelvinBot** [comment]
📣 @kuluruvineeth! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/231263719-a8a3b7d3-36a0-4d7f-8ad4-cebe1f560fa4.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by kuluruvineeth** [comment]
Contributor details
Your Expensify account email: kuluruvineeth8623@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0164ed01f32c785903

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
New room page does not auto focus in iOS. (new task page also suffer the same issue)

### What is the root cause of that problem?
I don't really have a deep explanation on why this issue happens. But here is what I found so far:

The autofocus issue only happens when we navigate to the page while a modal opens. When a modal opens and we do a navigation, the input will be auto focused, but also will be auto blurred for some reason (that's why we can see the validation error when we open the new room page). It doesn't happen with `BankAccountManualStep` because there is no modal there. It is also don't happen when we go to New chat page because we delayed the render of the input and the list and show it when `didScreenTransitionEnd` becomes true. However, the initial intention of using `didScreenTransitionEnd` is to fix animation related issue here https://github.com/Expensify/App/issues/2335, which I can't reproduce so I don't think we should use this solution. 

### What changes do you think we should make in order to solve the problem?


We previously already had a similar issue here https://github.com/Expensify/App/issues/15276 on web but I also didn't find the root cause really explain the cause deeply. The point of the solution is to navigate after a modal is completely closed. We can use `Modal.close()` and pass a function (which is the `navigate`) as the parameter that will be executed after a modal is completely closed.

https://github.com/Expensify/App/blob/f8acfae37e181c66ab9dbf232772bb293c9b159b/src/pages/home/sidebar/SidebarScreen/FloatingActionButtonAndPopover.js#L190

Note:
1. We should always delay the focus for Android (which we already did, no changes)
2. For iOS and Web, navigate after a modal is completely closed

To solve it globally, I think we can put every code of `navigate` inside `Modal.close`
https://github.com/Expensify/App/blob/f8acfae37e181c66ab9dbf232772bb293c9b159b/src/libs/Navigation/Navigation.js#L139-L172

**Comment by MelvinBot** [comment]
Looks like something related to `react-navigation` may have been mentioned in this issue discussion.

As a reminder, please make sure that all proposals are not workarounds and that any and all attempt to fix the issue holistically have been made before proceeding with a solution. Proposals to change our `DeprecatedCustomActions.js` files should not be accepted.

Feel free to drop a note in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q) with any questions.

**Comment by hellohublot** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Auto focus is broken on New room page

### What is the root cause of that problem?

In `PopoverMenu`, when we select an item, we will execute `this.selectItem -> this.props.onItemSelected -> hideCreateMenu`, when the Modal closing animation is executed, it will passively call `onModalHide -> resetFocusAndHideModal -> Navigation.navigate()`, we navigate to a new page after the Modal is completely hidden, this is indeed the right logic

However, on iOS, when we hide the Modal, `RCTModalHostViewManager.dismissModalHostView` will execute `dismissViewControllerAnimated`, which is a system method of iOS, which will call UIViewController for the upcoming `rootView(the rootView under the Modal).becomeFirstResponder()`, which will make other `UITextField` lose focus, will call `UITextField.resignFirstResponder`

### What changes do you think we should make in order to solve the problem?

So, we should wait for `dismissViewControllerAnimated` to execute before executing `onModalHide -> Navigation.navigate`

Since only iOS has an `onDismiss` callback, 
`react-native-modal` has not been maintained for many years, so I have to use `react-native-modal.patch`, I also hope you can fork this repository
https://github.com/react-native-modal/react-native-modal/blob/6624b5abb326470820f9fedec8b3ecdb35520869/src/modal.tsx#L684-L686
```diff
+   // We will handle `onModalHide`` in the `onDismiss` event (iOS only)
+   if (Platform.OS === 'ios') {
+       return;
+   }
+   this.props.onModalHide();

...


- return (React.createElement(Modal, Object.assign({ transparent: true, animationType: 'none', 
-    visible: this.state.isVisible, onRequestClose: onBackButtonPress }, otherProps),
+ return (React.createElement(Modal, Object.assign({ transparent: true, animationType: 'none', 
+    visible: this.state.isVisible, onRequestClose: onBackButtonPress, 
+    onDismiss: Platform.OS === 'ios' ? this.props.onModalHide : undefined }, otherProps)
```

**Comment by techievivek** [comment]
Based on @aimane-chnaif's comment [here](https://github.com/Expensify/App/issues/17479#issuecomment-1526046605), I am going to put this on HOLD. We can revisit the issue to see if this is reproducible or not once the PR from the linked issue is merged. Thanks

**Comment by johncschuster** [comment]
Thanks, @techievivek! Should we downgrade the priority label as well while we wait, or should we keep it on `Daily`?

**Comment by techievivek** [comment]
Good point, I am updating it to weekly for now. Thanks

**Comment by johncschuster** [comment]
@techievivek, are we still held on this one?

**Comment by techievivek** [comment]
@johncschuster Yeah, it looks the PR review is still in progress, so we will need to HOLD this for some more time.

**Comment by techievivek** [comment]
Not overdue, the linked PR is still in progress.

**Comment by bernhardoj** [comment]
I can't reproduce this issue anymore

**Comment by techievivek** [comment]
@bernhardoj This is still happening, I am on main(latest) and was able to reproduce this. Also, it looks like the linked [PR](https://github.com/Expensify/App/pull/19011) is still in progress.


https://github.com/Expensify/App/assets/35863227/6a5e4161-7fb4-483b-98e2-36791da70a58

**Comment by bernhardoj** [comment]
My bad. Just retested and the auto focus works only once and broken on the next open.

**Comment by bernhardoj** [comment]
My bad. Just retested and the auto focus works only once and broken on the next open.

**Comment by johncschuster** [comment]
Linking the PR in progress so it's easier to get a sense of our status:

https://github.com/Expensify/App/pull/19011

**Comment by techievivek** [comment]
The [PR](https://github.com/Expensify/App/pull/19011) on which we are holding this issue to is still being reviewed.

**Comment by techievivek** [comment]
The PR is expected to get merged in a day or two https://github.com/Expensify/App/pull/19011#issuecomment-1634521350, and then we should be able to make progress here. Thanks for understanding.

**Comment by techievivek** [comment]
There are some last-minute discussions happening on the linked PR https://github.com/Expensify/App/pull/19011#issuecomment-1650328856. However, we hope to have it merged within the next day or two. Once that happens, we can revisit the issue at hand.

**Comment by johncschuster** [comment]
Chill, Melv. We're still waiting on the above PR to be merged.

**Comment by techievivek** [comment]
Ok, the linked [PR](https://github.com/Expensify/App/pull/19011) is merged and deployed to PROD, so we should be able to move forward here now.

**Comment by johncschuster** [comment]
Great! Thanks for the update, @techievivek!

@thesahindia, how do you feel about the available proposals? Are we good to assign this one out or should we raise the bounty?

**Comment by techievivek** [comment]
Gentle bump @thesahindia ^

**Comment by situchan** [comment]
I can review this as I already know context as a bug reporter.

**Comment by techievivek** [comment]
Added @situchan as the C+ for this issue based on the comment https://github.com/Expensify/App/issues/17479#issuecomment-1687741510

**Comment by situchan** [comment]
@hellohublot I think [your patch](https://github.com/Expensify/App/pull/23994/files#diff-1c20fd843c83bf35eedb11b8ecd7c369381522da830c8becafa2467376c4d232) will fix this issue as well.
Can you test and confirm?

**Comment by techievivek** [comment]
Not overdue. @hellohublot can you please take a look at the above comments and respond? Thanks

**Comment by situchan** [comment]
@techievivek it's confirmed - https://github.com/Expensify/App/pull/23994#issuecomment-1690239398
Let's just wait for that PR to hit production.

**Comment by techievivek** [comment]
Awesome, can't we test this on staging to confirm if it fixes this bug or not?

**Comment by situchan** [comment]
> Awesome, can't we test this on staging to confirm if it fixes this bug or not?

Ah yes we can test on staging but only internally. External contributors don't have access to staging iOS build

**Comment by johncschuster** [comment]
@techievivek I'm a bit confused on where we stand on this one. Do we need to test this internally?

**Comment by johncschuster** [comment]
Thanks, @situchan! In that case, can we close this one out?

**Comment by situchan** [comment]
> Thanks, @situchan! In that case, can we close this one out?

yes, and I think I am eligible for reporting bonus

**Comment by techievivek** [comment]
Not overdue, we just need to issue the reporting bonus here.

**Comment by techievivek** [comment]
@johncschuster Can we issue the reporting bonus to @situchan and close this issue? Thanks

**Comment by johncschuster** [comment]
Sorry I missed this! Yep! I'll get the reporting bonus issued.

**Comment by situchan** [comment]
@johncschuster I already have accepted contract from $250 offer. Thanks

**Comment by techievivek** [comment]
@johncschuster Have we finished the payment here? Can I close the issue?

**Comment by johncschuster** [comment]
Yep! Payment has been issued! We're good to close this one out.
